src/main/java/com/google/sps/servlets/PacketParserServlet.java

```diff
@@ -0,0 +1,167 @@
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+import java.util.*; 
+import com.google.gson.Gson;
+
+import java.io.IOException;
+import java.io.*;
+
+/** Servlet that processes raw PCAP files within the repo. 
+   * We define MYIP to be the IP address of the local machine that was used to create the PCAP file, 
+     and OUTIP to be all of the IP addresses that machine connected with.
+   * allPCAP HashMap indexes all unique OUTIPs in a PCAP file to a PCAPdata object that stores needed attributes, 
+     including the number of connections between MYIP and OUTIP 
+   * We load datastore with all of the unique PCAPdata objects from our HashMap with a given entity tag.
+   * Hard-coded the source IP address (MYIP) -- need to find a way to retrieve MYIP (look at first packet, ask user, use WhatsMyIP?)
+*/
+@WebServlet("/PCAP-data")
+public class PacketParserServlet extends HttpServlet {
+  HashMap<String, PCAPdata> allPCAP = new HashMap<String, PCAPdata>();
+  DatastoreService datastore = DatastoreServiceFactory.getDatastoreService(); //creates database
+
+  //static final String FILENAME = "WEB-INF/files/traffic.pcap";
+  //static final String FILENAME = "WEB-INF/chargen-udp.pcap";
+  //static final String FILENAME = "WEB-INF/files/chargen-tcp.pcap";
+  static final String FILENAME = "WEB-INF/files/smallFlows.pcap";
+
+  static final String MYIP = "192.168.3.131"; // for smallFlows.pcap 
+  static final String ENTITY_TAG = "file_1";
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    parseRaw(FILENAME);
+    putDatastore(ENTITY_TAG);
+  }
+
+  /* Reads PCAP file (from file name) as a stream and puts unique connections into allPCAP HashMap. */
+  public void parseRaw(String file) throws IOException {
+    final InputStream stream = new FileInputStream(file);
+    final Pcap pcap = Pcap.openStream(stream);
+
+    pcap.loop(new PacketHandler() {
+      @Override
+      public boolean nextPacket(final Packet packet) throws IOException {
+      if(packet.hasProtocol(Protocol.IPv4)) {// is the first packet always (MYIP, OUTIP)?
+        IPPacket ip = (IPPacket) packet.getPacket(Protocol.IPv4);
+        String protocol = getProtocol(ip); 
+        String OUTIP = "";
+
+        Buffer buffer = ip.getPayload();
+        int size = buffer.getReadableBytes();
+            
+        //The IP addresses involved
+        String srcip = ip.getSourceIP();
+        String dstip = ip.getDestinationIP();
+
+        if (srcip == MYIP) {
```

Comment:
you are combining the processing in the parsing. I know it is efficient to do that while you are parsing but that is a trade-off worth since parsing is several orders of magnitude more time than looping the array again. So try to see if you can decouple this.